cleaner.go

```diff
@@ -0,0 +1,362 @@
+			cleaner.docIdsToRemove = append(cleaner.docIdsToRemove, doc.Id)
+		} else if len(toPull) > 0 {
+			// Note: (jam 2017-04-04) An observation, if it is legal to pull a token from one document,
+			// it is legal to pull it from all other documents. We could do bulk operations by using the
+			// union of all document ids and the union of all tokens to pull.
+			cleaner.stats.PulledTokenCount += len(toPull)
+			pull := bson.M{"$pullAll": bson.M{"txn-queue": toPull}}
+			pullChunk.Update(bson.M{"_id": doc.Id}, pull)
+			if pullCount >= maxBulkOps {
+				if err := flushPulls(); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	if err := flushPulls(); err != nil {
+		return err
+	}
+	// We've handled these tokens and documents
+	cleaner.tokensToLookup = cleaner.tokensToLookup[:0]
+	cleaner.docsToProcess = cleaner.docsToProcess[:0]
+	return nil
+}
+
+// flushRemoveQueue ensures that all pending removals are flushed to the database.
+func (cleaner *collectionCleaner) flushRemoveQueue() error {
+	if len(cleaner.docIdsToRemove) == 0 {
+		return nil
+	}
+	remover := newBulkRemover(cleaner.config.Source)
+	for _, docId := range cleaner.docIdsToRemove {
+		if err := remover.remove(docId); err != nil {
+			return fmt.Errorf("failed while removing document %v from %q", docId, cleaner.config.Source.Name)
+		}
+	}
+	if err := remover.flush(); err != nil {
+		return fmt.Errorf("failed while removing documents from %q", cleaner.config.Source.Name)
+	}
+	cleaner.stats.RemovedCount += remover.removed
+	logger.Debugf("flushing %d documents removed %d (%d total)",
+		len(cleaner.docIdsToRemove), remover.removed, cleaner.stats.RemovedCount)
+	cleaner.docIdsToRemove = cleaner.docIdsToRemove[:0]
+	return nil
+}
+
+// Cleanup iterates the collection and ensures that all documents no longer reference completed transactions.
+func (cleaner *collectionCleaner) Cleanup() error {
+	// walk through all the documents and remove what we can, update what we cannot
+	logger.Debugf("cleaning up completed documents from %q", cleaner.config.Source.Name)
+	// If we delete documents while we iterate, it can cause the iterator to miss documents.
+	// So we do multiple passes on the database to make sure we catch everything.
+	var doc txnDocument
+	for iterCount := 0; iterCount < maxIterCount; iterCount++ {
+		removedWhileIterating := false
+		query := cleaner.config.Source.Find(nil)
+		query.Batch(maxBatchDocs)
+		iter := query.Iter()
+		for iter.Next(&doc) {
+			if cleaner.massageId != nil {
+				doc.Id = cleaner.massageId(doc.Id)
+			}
+			if shouldFlush, err := cleaner.includeDoc(doc); err != nil {
+				return err
+			} else if shouldFlush {
+				if err := cleaner.processStashDocs(); err != nil {
+					return err
+				}
+				if len(cleaner.docIdsToRemove) > cleaner.config.MaxRemoveQueue {
+					if err := cleaner.flushRemoveQueue(); err != nil {
+						return err
+					}
+					removedWhileIterating = true
+				}
+			}
+		}
+		if err := iter.Close(); err != nil {
```

Comment:
I wonder if this should go towards the bottom of the loop so that any queued removals get flushed even if the iterator has problems.